nym-node-status-api/nym-node-status-api/src/http/state.rs

```
@@ -158,17 +168,166 @@ impl HttpCache {
 
                 if gateways.is_empty() {
-                    tracing::warn!("Database contains 0 gateways");
+                    tracing::warn!("Database: gateway list is empty");
+                } else {
+                    self.upsert_gateway_list(gateways.clone()).await;
                 }
 
                 gateways
             }
         }
     }
 
+    pub async fn upsert_dvpn_gateway_list(
+        &self,
+        new_gateway_list: Vec<DVpnGateway>,
+    ) -> Entry<String, Arc<RwLock<Vec<DVpnGateway>>>> {
+        self.dvpn_gateways
+            .entry_by_ref(DVPN_GATEWAYS_LIST_KEY)
+            .and_upsert_with(|maybe_entry| async {
+                if let Some(entry) = maybe_entry {
+                    let v = entry.into_value();
+                    let mut guard = v.write().await;
+                    *guard = new_gateway_list;
+                    v.clone()
+                } else {
+                    Arc::new(RwLock::new(new_gateway_list))
+                }
+            })
+            .await
+    }
+
+    pub async fn get_dvpn_gateway_list(
+        &self,
+        db: &DbPool,
+        min_node_version: &Version,
+    ) -> Vec<DVpnGateway> {
+        match self.dvpn_gateways.get(DVPN_GATEWAYS_LIST_KEY).await {
+            Some(guard) => {
+                tracing::trace!("Fetching from cache...");
+                let read_lock = guard.read().await;
+                read_lock.clone()
+            }
+            None => {
+                tracing::trace!("No gateways (dVPN) in cache, refreshing from DB...");
+
+                let gateways = self.get_gateway_list(db).await;
+
+                let started_with = gateways.len();
+                let Ok(skimmed_nodes) = crate::db::queries::get_described_bonded_nym_nodes(db)
+                    .await
+                    .map(|records| {
+                        records
+                            .into_iter()
+                            .filter_map(|dto| SkimmedNode::try_from(dto).ok())
```

we're kinda silently ignoring nodes that are seemingly invalid. at the very least we should log them to know about it

---

true, added a log